photon-client/src/main/scala/com/linkedin/photon/ml/data/DataValidators.scala

```diff
@@ -259,26 +259,26 @@ object DataValidators extends Logging {
       perSampleValidators: List[(((Row, String) => Boolean), InputColumnsNames.Value, String)],
       inputColumnsNames: InputColumnsNames,
       featureSectionKeys: Set[FeatureShardId]): Seq[String] = {
-
-    perSampleValidators
-      .flatMap { case (validator, columnName, msg) =>
-        val rows = dataSet.collect()
-        rows.flatMap { r =>
-          if (columnName.equals(InputColumnsNames.FEATURES_DEFAULT)) {
-            featureSectionKeys.map(features => (validator(r, features), msg))
+    
+    val columns = dataSet.columns
+    dataSet.rdd.flatMap { r =>
+      perSampleValidators.flatMap { case (validator, columnName, msg) =>
+        if (columnName.equals(InputColumnsNames.FEATURES_DEFAULT)) {
+          featureSectionKeys.map(features => (validator(r, features), msg))
+        } else {
+          val result = if (columns.contains(inputColumnsNames(columnName))) {
+            (validator(r, inputColumnsNames(columnName)), msg)
           } else {
-            val result = if (dataSet.columns.contains(inputColumnsNames(columnName))) {
-              (validator(r, inputColumnsNames(columnName)), msg)
-            } else {
-              (true, "")
-            }
-            
-            Seq(result)
+            (true, "")
           }
+
+          Seq(result)
         }
       }
-      .filterNot(_._1)
-      .map(_._2)
+    }
+    .collect()
+    .filterNot(_._1)
```

Comment:
I'd like @joshvfleming to confirm, but I believe filtering first and then collecting would be more efficient.

Reply:
@ashelkovnykov is right -- as it stands, what this will do is collect validation results for all data points in the Driver and then filter them locally. If we filter first, we pull only the validation results we're interested in to the Driver.